rust/lance-encoding/src/encodings/physical/dictionary.rs

```diff
@@ -0,0 +1,410 @@
+            let items_decoder: Arc<dyn PrimitivePageDecoder> = Arc::from(items_page_decoder.await?);
+
+            let mut primitive_wrapper = PrimitiveFieldDecoder::new_from_data(
+                items_decoder.clone(),
+                DataType::Utf8,
+                copy_size,
+            );
+
+            // Decode all items
+            let drained_task = primitive_wrapper.drain(copy_size)?;
+            let items_decode_task = drained_task.task;
+            let mut decoded_dict = items_decode_task.decode()?;
+
+            // For the case where the dictionary is encoded as a single null element
+            // Reset it to be an empty string array
+            if decoded_dict.len() == 1 && decoded_dict.is_null(0) {
+                decoded_dict = Arc::new(StringArray::from(Vec::<Option<String>>::new()));
+            }
+
+            let indices_decoder: Box<dyn PrimitivePageDecoder> = indices_page_decoder.await?;
+
+            Ok(Box::new(DictionaryPageDecoder {
+                decoded_dict,
+                indices_decoder,
+                items_decoder,
+            }) as Box<dyn PrimitivePageDecoder>)
+        })
+        .map(|join_handle| join_handle.unwrap())
+        .boxed()
+    }
+}
+
+struct DictionaryPageDecoder {
+    decoded_dict: Arc<dyn Array>,
+    indices_decoder: Box<dyn PrimitivePageDecoder>,
+    items_decoder: Arc<dyn PrimitivePageDecoder>,
+}
+
+impl PrimitivePageDecoder for DictionaryPageDecoder {
+    fn decode(
+        &self,
+        rows_to_skip: u64,
+        num_rows: u64,
+        all_null: &mut bool,
+    ) -> Result<Vec<BytesMut>> {
+        // Decode the indices
+        let indices_buffers = self
+            .indices_decoder
+            .decode(rows_to_skip, num_rows, all_null)?;
+
+        let indices_array =
+            new_primitive_array::<UInt8Type>(indices_buffers.clone(), num_rows, &DataType::UInt8);
+
+        let indices_array = indices_array.as_primitive::<UInt8Type>().clone();
+        let dictionary = self.decoded_dict.clone();
+
+        let adjusted_indices: UInt8Array = indices_array
+            .iter()
+            .map(|x| match x {
+                Some(0) => None,
+                Some(x) => Some(x - 1),
+                None => None,
+            })
+            .collect();
+
+        // Build dictionary array using indices and items
+        let dict_array =
+            DictionaryArray::<UInt8Type>::try_new(adjusted_indices, dictionary).unwrap();
+        let string_array = arrow_cast::cast(&dict_array, &DataType::Utf8).unwrap();
+        let string_array = string_array.as_any().downcast_ref::<StringArray>().unwrap();
+
+        // This is not ideal, since we go from DictionaryArray -> StringArray -> buffers (BytesMut)
+        // and later in primitive_array_from_buffers() we will go from buffers -> StringArray again.
+        // Creating the BytesMut is an unnecessary copy.
+        // But it is the best we can do in the current structure
```

Comment:
Is this comment meant to apply to the `null_buffer`?  I think it might be more applicable for `bytes_buffer`?

Reply:
It was meant to apply to all 3 buffers (nulls, offsets, bytes)
I have updated the comment